alacritty/src/renderer/builtin_font.rs

```diff
@@ -0,0 +1,700 @@
+    }
+
+    /// Gives bounds for vertical straight line on `y` with `stroke_size`.
+    fn v_line_bounds(&self, x: f32, stroke_size: usize) -> (f32, f32) {
+        let start_x = cmp::max((x - stroke_size as f32 / 2.) as i32, 0) as f32;
+        let end_x = cmp::min((x + stroke_size as f32 / 2.) as i32, self.width as i32) as f32;
+
+        (start_x, end_x)
+    }
+
+    /// Draws a horizontal straight line from (`x`, `y`) of `size` with the given `stroke_size`.
+    fn draw_h_line(&mut self, x: f32, y: f32, size: f32, stroke_size: usize) {
+        let (start_y, end_y) = self.h_line_bounds(y, stroke_size);
+        self.draw_rect(x, start_y as f32, size, (end_y - start_y) as f32, COLOR_FILL);
+    }
+
+    /// Draws a vertical straight line from (`x`, `y`) of `size` with the given `stroke_size`.
+    fn draw_v_line(&mut self, x: f32, y: f32, size: f32, stroke_size: usize) {
+        let (start_x, end_x) = self.v_line_bounds(x, stroke_size);
+        self.draw_rect(start_x as f32, y, (end_x - start_x) as f32, size, COLOR_FILL);
+    }
+
+    /// Draws a rect from the (`x`, `y`) of the given `width` and `height` using `color`.
+    fn draw_rect(&mut self, x: f32, y: f32, width: f32, height: f32, color: Pixel) {
+        let start_x = x as usize;
+        let end_x = cmp::min((x + width) as usize, self.width);
+        let start_y = y as usize;
+        let end_y = cmp::min((y + height) as usize, self.height);
+        for y in start_y..end_y {
+            let y = y * self.width;
+            self.buffer[start_x + y..end_x + y].fill(color);
+        }
+    }
+
+    /// Put pixel into buffer with the given color if the color is brighter than the one buffer
+    /// already has in place.
+    #[inline]
+    fn put_pixel(&mut self, x: f32, y: f32, color: Pixel) {
+        let x = (x as usize).clamp(0, self.width - 1);
+        let y = (y as usize).clamp(0, self.height - 1);
+        let index = x as usize + y as usize * self.width;
+        if color._r > self.buffer[index]._r {
+            self.buffer[index] = color;
+        }
+    }
+
+    /// Xiaolin Wu's line drawing from (`from_x`, `from_y`) to (`to_x`, `to_y`).
+    fn draw_line(&mut self, mut from_x: f32, mut from_y: f32, mut to_x: f32, mut to_y: f32) {
+        let steep = (to_y - from_y).abs() > (to_x - from_x).abs();
+        if steep {
+            mem::swap(&mut from_x, &mut from_y);
+            mem::swap(&mut to_x, &mut to_y);
+        }
+        if from_x > to_x {
+            mem::swap(&mut from_x, &mut to_x);
+            mem::swap(&mut from_y, &mut to_y);
+        }
+
+        let delta_x = to_x - from_x;
+        let delta_y = to_y - from_y;
+        let gradient = if delta_x.abs() <= f32::EPSILON { 1. } else { delta_y / delta_x };
+
+        let x_end = f32::round(from_x);
+        let y_end = from_y + gradient * (x_end - from_x);
+        let x_gap = (from_x + 0.5).fract();
+
+        let xpx11 = f32::floor(x_end);
+        let ypx11 = f32::floor(y_end);
```

Comment:
I'd assume these variable names come from the wikipedia entry for the algorithm? In which case it's actually `xpxl1`/`ypxl1`.